% Preamble
\documentclass[12pt,letterpaper]{article}
\title{Supplementary Material}
\author{Yihang Cheng}
\date{\today}
% Packages
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage{multirow}
\usepackage{array}
\usepackage[colorlinks=true, linkcolor=blue, urlcolor=blue, citecolor=blue]{hyperref}
% \usepackage{listings}
\usepackage{xcolor}
\usepackage{threeparttable}
\usepackage{float}
\usepackage{algorithm}
\usepackage{algorithmic}
% \usepackage[style=authoryear,backend=bibtex]{biblatex}
% \addbibresource{SupplymentaryMaterialReference.bib}
% Document
\begin{document}
\maketitle
\newpage
\tableofcontents
\newpage
\section{Software tools developed in this study}\label{sec:software-develope}
The analysis algorithms and methodologies employed in this paper were integrated into two tool sets:

\begin{itemize}
    \item  \href{https://github.com/hcyvan/epiLungCancer/tree/main/methytools}{\textit{methytools}}
    \item \href{https://github.com/hcyvan/pattools}{\textit{pattools}}
\end{itemize}

\subsection{methytools}\label{sec:software-develope-methytools}

methytools is a toolkit developed for processing BS-seq methylation data.

\subsubsection{mcomppost percentile}

\emph{mcomppost percentile} can be used to calculate the percentile of each CpGs and add percentile columns to the DMCs files.

\begin{verbatim}
mcomppost percentile -i ./dmc.Rest.vs.CTL.txt \
                    -m ./merge.d3.all.bed.gz \
                    -t ./sample.CTL.txt \
                     -o ./dmc.Rest.vs.CTL.percentile.txt
\end{verbatim}

\subsection{pattools}\label{sec:software-develope-pattools}

pattools is a toolkit based on PAT format\cite{Loyfer2024.05.08.593132} data, inheriting both existing
methylation algorithms and incorporating new ones.
    {pattools} is currently under development, and the algorithms from this toolkit are primarily utilized
in this study. The software will be introduced in a forthcoming paper.

\subsubsection{pattools vector}

\section{Selection of subtype-specific DMCs}\label{sec:selection-of-dmcs}

\subsection{One-vs-Rest}

We employed the One-vs-Rest strategy to identify specific differentially methylated CpGs (DMCs) across various
subtypes of lung cancer, using MOABS\cite{sun_moabs_2014} for preliminary screening. The target group and Background
group are shown in the following table \ref*{tab:grouping}.

\begin{table}[!h]
    \begin{center}
        \caption{ Grouping of One-vs-Rest }
        \begin{tabular}{|l|l|}
            \hline
            Target & Background            \\
            \hline
            CTL    & LUAD, LUSC, LCC, SCLC \\
            LUAD   & CTL, LUSC, LCC, SCLC  \\
            LUSC   & CTL, LUAD, LCC, SCLC  \\
            LCC    & CTL, LUAD, LUSC, SCLC \\
            SCLC   & CTL, LUAD, LUSC, LCC  \\
            \hline
        \end{tabular}
    \end{center}
    \label{tab:grouping}
\end{table}

\subsection{MOABS mcomp}

Using the default parameters of MOABS, we obtained a large number of DMCs in each comparison group (Table \ref{tab:table_dmc_ount}).
Subsequently, duplicated differentially methylated CpG sites (DMCs) across multiple groups were excluded, leaving
only those specific to CTL, LUAD, LUSC, LCC, and SCLC.

\begin{table}[!h]
    \begin{center}
        \caption{The count of DMCs find by MOABS}
        \begin{threeparttable}
            \begin{tabular}{|l|p{0.2\textwidth}|p{0.2\textwidth}|}
                \hline
                \multirow{2}{*}{Group} & \multicolumn{2}{c|}{Filtered DMCs\tnote{1}}           \\
                \cline{2-3}            & Befor                                       & After   \\
                \hline
                CTL vs. Rest           & 584239                                      & 139129  \\
                LUAD vs. Rest          & 95999                                       & 30220   \\
                LUSC vs. Rest          & 427525                                      & 318345  \\
                LCC vs. Rest           & 1361568                                     & 1068507 \\
                SCLC vs. Rest          & 3022977                                     & 2173620 \\
                \hline
            \end{tabular}

            \begin{tablenotes}
                \item[1] Exclude the following DMCs:
                1. Those that appear in more than one group.
                2. Those located on the sex chromosomes
            \end{tablenotes}

        \end{threeparttable}
    \end{center}
    \label{tab:table_dmc_ount}
\end{table}

\subsection{\textit{p}-th percentile algorithm}

A more stringent strategy was then used for further screening of DMCs. We evaluate the
\textit{p}-th percentile of low-methylated group and the (\textit{100 - p})-th percentile
of high-methylated group. In hypermethylated DMCs, the \textit{p}-th percentile of the target
group and the (\textit{100 - p})-th percentile of the background are calculated; conversely,
in hypomethylated DMCs, this relationship is reversed. If the \textit{p}-th percentile of the
low-methylated group is less than or equal to (\textit{100-p})-th percentile of the high-methylated
group, we assert that this DMCs adheres to our screening criteria (Figure \ref*{fig:dmcPC}).

\begin{figure}{ht}
    \centering
    \includegraphics[width=0.8\linewidth]{./img/dmcPercentileCut.png}
    \caption{The 85-th percentile of target and background samples in a hypomethylated DMCs}
    \label{fig:dmcPC}
\end{figure}

Utilizing \emph{mcomppost}, we conducted an analysis of the DMCs within each group. As
the \textit{p}-th percentile increased, the number of retained DMCs decreased significantly
(Figure \ref*{fig:dmcPK}). By selecting an appropriate percentile (\textit{p}) threshold,
we can identify DMCs that more stringently adhere to
group specificity.


\begin{figure}[htbp]
    \centering
    \includegraphics[width=1\linewidth]{./img/dmcPercentileVsKeepProp}
    \caption{The remaining DMC propertion after filtering with different percentiles}
    \label{fig:dmcPK}
\end{figure}

\section{Deconvolution}\label{sec:deconvolution}
Three deconvolution algorithms (lofer\cite{Loyfer2024.05.08.593132}, moss \cite{moss_comprehensive_2018}
and sun\cite{sun_plasma_2015}) have been utilized to perform the deconvolution of the tissue samples.

To mitigate the interference caused by the presence of numerous cell/tissue types in
lung tissue deconvolution, we manually selected specific cell/tissue types that are potentially
implicated in lung cancer for this analysis (Table \ref*{tab:deconvolution}). For clarification
on the cell/tissue types, \textit{pattools deconv-helper} can be used to view the corresponding
cell/tissue types for each category.

\begin{table}[htbp]
    \begin{center}
        \caption{Cell/Tissue types used by each algorithm }
        \begin{tabular}{|l|l|l|l|}
            \hline
            \textbf{Method}                         & \textbf{Refined group}                      & \textbf{Group}           & \textbf{Note} \\
            \hline
            \multirow{6}{*}{Sun \textit{et al.}}    & Lung cells                                  & Lungs                    &               \\
            \cline{2-4}                             & Neural and endocrine cells                  & Brain                    &               \\
            \cline{2-4}                             & \multirow{3}{*}{Immune cells}               & T-cells                  &               \\
            \cline{3-4}                             &                                             & B-cells                  &               \\
            \cline{3-4}                             &                                             & Neutrophils              &               \\
            \cline{2-4}                             & Others                                      & Heart                    &               \\
            \hline
            \multirow{6}{*}{Moss \textit{et al.}}   & Lung cells                                  & LungCells                &               \\
            \cline{2-4}                             & \multirow{2}{*}{Neural and endocrine cells} & CorticalNeurons          &               \\
            \cline{3-4}                             &                                             & PancreaticBetaCells      &               \\
            \cline{2-4}                             & \multirow{6}{*}{Immune cells}               & Cd4tCells                &               \\
            \cline{3-4}                             &                                             & Cd8tCells                &               \\
            \cline{3-4}                             &                                             & BCells                   &               \\
            \cline{3-4}                             &                                             & NkCells                  &               \\
            \cline{3-4}                             &                                             & Monocytes                &               \\
            \cline{3-4}                             &                                             & Neutrophils              &               \\
            \cline{2-4}                             & \multirow{2}{*}{Others}                     & VascularEndothelialCells &               \\
            \cline{3-4}                             &                                             & LeftAtrium               &               \\
            \hline
            \multirow{6}{*}{Loyfer \textit{et al.}} & \multirow{2}{*}{Lung cells}                 & Lung-Ep-Alveo            &               \\
            \cline{3-4}                             &                                             & Lung-Ep-Bron             &               \\
            \cline{2-4}                             & \multirow{5}{*}{Neural and endocrine cells} & Neuron                   &               \\
            \cline{3-4}                             &                                             & Oligodend                &               \\
            \cline{3-4}                             &                                             & Pancreas-Alpha           &               \\
            \cline{3-4}                             &                                             & Pancreas-Beta            &               \\
            \cline{3-4}                             &                                             & Pancreas-Delta           &               \\
            \cline{2-4}                             & \multirow{5}{*}{Immune cells}               & Blood-T                  &               \\
            \cline{3-4}                             &                                             & Blood-B                  &               \\
            \cline{3-4}                             &                                             & Blood-NK                 &               \\
            \cline{3-4}                             &                                             & Blood-Mono+Macro         &               \\
            \cline{3-4}                             &                                             & Blood-Granul             &               \\
            \cline{2-4}                             & \multirow{2}{*}{Others}                     & Endothel                 &               \\
            \cline{3-4}                             &                                             & Heart-Fibro              &               \\
            \cline{3-4}                             &                                             & Head-Neck-Ep             &               \\
            \hline
        \end{tabular}
    \end{center}
    \label{tab:deconvolution}
\end{table}

For BS-seq data, wgbstools\cite{Loyfer2024.05.08.593132} were initially employed to convert the
aligned BAM format data into PAT format. Subsequently, the methylation metrics required by the corresponding
algorithm were extracted from the PAT format for deconvolution analysis. The algorithm utilized in this
study for deconvolution was NNLS. All three deconvolution algorithms discussed in this article were
implemented using \textit{pattools deconv}.

This code performs the deconvolution of a sample using algorithm developed by Sun \textit{et al.} with \textit{pattools deconv}.

\begin{verbatim}
pattools deconv -m sun -g hg38 \
                -c /path/to/references/hg38/CpG.bed.gz \
                -p sample.pat.gz -o sample.sun.tsv \
                --include Lungs Heart Brain T-cells B-cells Neutrophils
\end{verbatim}

\section{Genes associated with lung cancer subtypes}\label{sec:genes}

\subsection{Genes summary}\label{sec:genes-summary}

There is no simple one-to-one relationship between lung cancer subtypes and cells of origin. Subtypes are
influenced by both the originating cells and the driver mutations\cite{ferone_cells_2020}.

\begin{table}[htbp]
    \begin{center}
        \caption{ Marker genes of interest}
        \begin{tabular}{|l|l|l|}
            \hline
            \textbf{Subtype}      & \textbf{Gene} & \textbf{Note}                              \\
            \hline
            LUAD                  & NKX2-1        & upregulated in LUAD                        \\
            \hline
            \multirow{2}{*}{LUSC} & SOX2          & upregulated in LUSC                        \\
            \cline{2-3}           & P63           & basal cell marker                          \\
            \hline
            \multirow{4}{*}{SCLC} & ASCL1         & \multirow{4}{*}{ref\cite{baine_sclc_2020}} \\
            \cline{2-2}           & NEUROD1       &                                            \\
            \cline{2-2}           & POU2F3        &                                            \\
            \cline{2-2}           & YAP1          &                                            \\
            \hline
        \end{tabular}
    \end{center}
    \label{tab:gene-marker}
\end{table}
\subsection{Extraction of methylation levels in gene regulatory regions}\label{sec:gene-methylation}

\section{Methylation vector}\label{sec:vector}

We represent methylated CpGs as 1 and unmethylated CpGs as 0, utilizing a 
0-1 array to indicate the methylation status of consecutive CpGs within the 
genome. This vector, composed of n binary values, is referred to as a 
methylation vector (MV).

\subsection{Methylation vector analysis algorithm}\label{sec:ssmv}

Identification of subtype-specific MVs involves the following three steps:

\begin{itemize}
    \item  Vectorization
    \item  Clustering
    \item  Seperating
\end{itemize}

\subsubsection{Vectorization}

This step was implemented using the \textit{pattools mv-vectorization} command. For each sample, 
a sliding window of length w CpGs is applied with a step size of 1 CpG site, resulting in a series 
of overlapping windows where adjacent windows share w-1 CpGs. Within each window, methylated CpGs 
are represented as 1 and unmethylated CpGs as 0, forming a 0-1 vector. Vectors with dimensions less 
than w are discarded.

\subsubsection{Clustering}

This step was implemented using the \textit{pattools mv-clustering} command. Merge MVs from all 
samples of CTL, LUAD, LUSC, LCC, and SCLC, and label each MV according to its sample and sample 
group of origin within each window. Apply the multiple repeats and equal spacing clustering (MRESC) 
algorithm to cluster the MVs in each window. 

Alternatively, existing clustering algorithms such as 
DBSCAN\cite{dbscanschubert}, HDBSCAN\cite{mcinnes2017hdbscan}, or KMeans\cite{kmeanskrishna} 
may be used in place of MRESC for this step.

\subsubsection{Seperating}

This step was implemented using the \textit{pattools mv-separating} command.

\subsection{MRESC}

In comparison to conventional clustering methods, the clustering of MVs exhibits 2 distinct 
characteristics:

\begin{itemize}
    \item  A limited number of MVs types exist within a fixed window, with a significant number 
    of these vectors recurring frequently. These identical MVs represent the same methylation 
    motif. In a window containing $w$ CpGs, there are $w^2$ possible methylation motifs. Each 
    motif may be observed multiple times. 
    \item  Methylation motifs are distributed equidistantly, forming a grid-like pattern. For 
    example, the manhattan distance between the MVs (0,0,0,0) and (1,0,0,0) is 1, and similarly,
    the distance between (0,0,0,0) and (0,1,0,0) is also 1.
\end{itemize}

\subsubsection{The distance between MVs}

In the representation of MVs, methylated CpGs are denoted by 1, while unmethylated CpGs are denoted 
by 0. In the representation of methylation motifs, 'C' represents methylated CpGs, and 'T' 
represents unmethylated CpGs. The distances between two distinct motifs are calculated and then 
sorted from smallest to largest. The distance between a given motif and itself is marked as 0. The 
distance between two adjacent moitfs is marked as 1. The distance between two moitfs separated by 
one interval is marked as 2, and so forth. In a window containing $w$ CpGs (Figure \ref*{fig:MRESC}), 
each methylation motif is adjacent to $w$ neighboring motifs, with a manhattan distance of 1.

\begin{figure}[ht]
    \centering
    \includegraphics[width=0.8\linewidth]{./img/mrescDist.png}
    \caption{
        All possible methylation motifs within a window containing 4 CpGs and the 
        distances between them. As the number in the heatmap increases, the distance 
        between the motifs becomes greater
    }
    \label{fig:MRESC}
\end{figure}

\subsubsection{Pseudocode}

Here is the pseudo code of MRESC (Algorithm \ref{alg:MRESC}). There are several key 
functions in this algorithm.

\begin{itemize}
    \item  \textbf{countVectorsInEachMotif:} Count the number of methylation vectors representing each 
    distinct methylation motif.
    \item  \textbf{findNeighborsIndex:} Find the neighboring motifs of a given motif
    \item  \textbf{findNeighborMax:} Identify the maximum MVs counts among the neighboring motifs of a given motif
    \item  \textbf{markCluster:} Assign a cluster label to each motif
    \item  \textbf{assignClusterLabel2Vectors:} Assign a cluster label to each MV
\end{itemize}

\begin{algorithm}[ht]
    \caption{\textbf{MRESC}}
    \label{alg:MRESC}
    \begin{algorithmic}
        \STATE \textbf{Function doMRESC}(vectors, w)
        \STATE \ \ \ \ counts $\gets$ \textbf{countVectorsInEachMotif}(vectors)
        \STATE \ \ \ \ clusters $\gets$ \textbf{[0]} * w
        \STATE \ \ \ \ \textbf{For} $i \gets 0$ \textbf{to} w - 1
        \STATE \ \ \ \ \ \ \ \ \textbf{For} $j$ \textbf{in} \textbf{findNeighborsIndex}(i)
        \STATE \ \ \ \ \ \ \ \ \ \ \ \ \textbf{If} counts[i] $<$ counts[j] \textbf{then}
        \STATE \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \textbf{Continue}
        \STATE \ \ \ \ \ \ \ \ \textbf{EndIf}
        \STATE \ \ \ \ \ \ \ \ max\_nb $\gets$ \textbf{findNeighborMax}(j, counts)
        \STATE \ \ \ \ \ \ \ \ \textbf{If} counts[i] $\geq$ max\_nb \textbf{then}
        \STATE \ \ \ \ \ \ \ \ \ \ \ \ \textbf{markCluster}(i, j, clusters)
        \STATE \ \ \ \ \ \ \ \ \textbf{EndIf}
        \STATE \ \ \ \ \textbf{EndFor}
        \STATE \ \ \ \ labels $\gets$ \textbf{assignClusterLabel2Vectors}(vectors, clusters)
        \STATE \ \ \ \ \textbf{Return} labels
        \end{algorithmic}
\end{algorithm}

\section{Abbreviations}\label{sec:abbr}

\begin{itemize}
    \item  \textbf{BS-seq}: Bisulfite sequencing
    \item  \textbf{NNLS}: non-negative least squares linear regression
    \item  \textbf{MVs}: Methylation vectors
\end{itemize}

\section{Code availability}\label{sec:code}

The analysis code of this artile a available in Github repository
\href{https://github.com/hcyvan/epiLungCancer}{epiLungCancer}. The sub project of epiLungCancer,
\href{https://github.com/hcyvan/epiLungCancer/tree/main/methytools}{\textit{methytools}},  was developed
to process BS-seq methylation data. Additionally, some of the analyses in this paper utilize code
from \href{https://github.com/hcyvan/pattools}{pattools}, which will be discussed in detail in
separate publications.


% \printbibliography
\bibliographystyle{abbrv}
\bibliography{SupplymentaryMaterialReference}
\end{document}